Inherit doc comments from default implementations of inherited protocols

ahoppen opened the pull request:
For example when we have

```swift
protocol BaseProtocol {}

extension BaseProtocol {
  /// Say hello
  func hello() { }
}
struct MyStruct: BaseProtocol {
  func hello() {}
}
```

Then `MyStruct.hello` should inherit the doc comment from the implementation of `BaseProtocol.hello`. Currently no doc comment is associated with `MyStruct.hello`.

rdar://126240546


---

The changes in offsets in the test cases is because `init(rawValue:)` now inherits the doc comment of `RawRepresentable`, which shifts the lines in generated interfaces around.

bnbarham commented on test/SourceKit/CursorInfo/cursor_inherit_doc_from_protocol_default_impl.swift:
```diff
@@ -0,0 +1,34 @@
+protocol BaseProtocol {}
```
Could add a test for the case where there is both a requirement and a default

ahoppen commented:
@swift-ci Please smoke test

hamishknight commented on lib/AST/DocComment.cpp:
```diff
@@ -426,24 +426,24 @@ class CommentProviderFinder final {
-    // Only applies to protocol extension member.
-    auto *protocol = VD->getDeclContext()->getExtendedProtocolDecl();
-    if (!protocol)
+    NominalTypeDecl *nominalType =
+        dyn_cast_or_null<NominalTypeDecl>(VD->getDeclContext()->getAsDecl());
```
What if `VD` is defined in an extension? Do we want to support that?

hamishknight replied:
That only handles protocol extensions, right? I was thinking of e.g a struct extension

hamishknight commented on lib/AST/DocComment.cpp:
```diff
@@ -426,24 +426,24 @@ class CommentProviderFinder final {
-                              DeclNameRef(VD->getName()),
-                              VD->getLoc(), NLOptions::NL_ProtocolMembers,
-                              members);
+    nominalType->lookupQualified(nominalType, DeclNameRef(VD->getName()),
+                                 VD->getLoc(), NLOptions::NL_ProtocolMembers,
+                                 members);
```
I guess this was also an issue with the previous code, but what if you have a decl that matches by name, but not by type? e.g something like:

```swift
protocol P {}
extension P {
  /// hello
  func foo(_ x: Int) {}
}

struct S: P {
  func foo(_ x: String) {}
}

func bar(_ x: S) {
  x.foo("")
}
```

IMO it could be unexpected to inherit the doc comment in that case.

ahoppen replied:
Oh, good idea. Added a test case based on the return type.

ahoppen commented on test/SymbolGraph/Relationships/Synthesized/RemoteInheritedDocs.swift:
```diff
@@ -53,9 +48,6 @@ import RemoteP
 // but `otherFunc` does. Regardless, each one needs a `sourceOrigin` field connecting it to
 // upstream.
 
-// `RemoteP.P` also has an extension with a default implementation for `extraFunc` that does have
-// docs, but overriding it here should prevent those from appearing
-
```
@QuietMisdreavus This comments seems to imply the exact opposite of what I want to achieve in this PR. Is there any reason why we didn’t want to inherit documentation from extensions?

For context, the concrete example I want to fix with this PR is that this implementation of `_ArrayProtocol.filter` (https://github.com/apple/swift/blob/c531f2914fda0b84d8af091a0081799101f76d9b/stdlib/public/core/ArrayType.swift#L73-L78) should inherit the documentation from the documentation of `Sequence.filter`, which is implemented in an extension (https://github.com/apple/swift/blob/5df0f053d95c1143ac34ac88faafdfa3902a3224/stdlib/public/core/Sequence.swift#L717-L739)

QuietMisdreavus replied:
cc @franklinsch (and possibly also @bitjammer for the underlying discussion of inheriting docs in the first place)

I think it should respect the presence of the `-skip-inherited-docs` flag, but otherwise i don't remember off the top of my head.

ahoppen replied:
I think we already respect the `-skip-inherited-docs` flag because we’re not calling into `getDocCommentProvidingDecl` if it is set and this PR only modifies `getDocCommentProvidingDecl`

https://github.com/apple/swift/blob/9fdb0e79e2dab72fdc01f8fa50d9cc249a48a5a0/lib/SymbolGraphGen/Symbol.cpp#L359-L365

hamishknight commented on lib/AST/DocComment.cpp:
```diff
@@ -426,24 +426,44 @@ class CommentProviderFinder final {
-                              DeclNameRef(VD->getName()),
-                              VD->getLoc(), NLOptions::NL_ProtocolMembers,
-                              members);
+    nominalType->lookupQualified(nominalType, DeclNameRef(VD->getName()),
+                                 VD->getLoc(), NLOptions::NL_ProtocolMembers,
+                                 members);
```
Thought of a couple more edge cases :)

1. static mismatch

```swift
protocol P {}
extension P {
  /// hello
  static func foo() {}
}
struct S: P {
  func foo() {}
}
func bar(_ x: S) {
  x.foo()
}
```

2. enum case matching a function (if we want this to work it would need to be a static function)

```swift
protocol P {}
extension P {
  /// hello
  func foo(_ x: Int) -> E { .foo(0) }
}

enum E: P {
  case foo(Int)
}

func bar(_ x: E) {
  E.foo(0)
}
```

3. this is quite silly, but technically this would match:

```swift
class C {
  /// hello
  struct foo {}
}
class D: C {
  var foo: C.foo.Type { C.foo.self }
}

func bar(_ x: D) {
  let y = x.foo
}
```

ahoppen replied:
Good cases, thanks. Case 3 does inherit but I decided to not care and not add a test case for it. If it broke, I don’t think we would care.

hamishknight commented on lib/AST/DocComment.cpp:
```diff
@@ -426,24 +426,44 @@ class CommentProviderFinder final {
+    if (!vdComparisonTy) {
+      return std::nullopt;
+    }
+    if (auto fnTy = vdComparisonTy->getAs<AnyFunctionType>()) {
+      // Strip off the 'Self' parameter.
+      vdComparisonTy = fnTy->getResult();
+    }
```
Wonder if we ought to have a utility for removing the curried self, could be used in a couple of other places:

https://github.com/apple/swift/blob/15fd443be60cfb52852c853c4fa0d6a1622ada10/lib/Sema/TypeCheckProtocol.cpp#L4894-L4896

https://github.com/apple/swift/blob/15fd443be60cfb52852c853c4fa0d6a1622ada10/lib/Sema/AssociatedTypeInference.cpp#L1990-L1993

https://github.com/apple/swift/blob/15fd443be60cfb52852c853c4fa0d6a1622ada10/lib/AST/Decl.cpp#L4937-L4939

ahoppen commented on lib/AST/DocComment.cpp:
```diff
@@ -426,24 +426,44 @@ class CommentProviderFinder final {
+        // Strip off the 'Self' parameter.
+        memberComparisonTy = fnTy->getResult();
+      }
+      if (!memberComparisonTy->matches(vdComparisonTy, TypeMatchFlags::AllowOverride)) {
         continue;
```
Oh, yes. I thought that I had this case covered with the `throws` case but apparently `throws` works either way 🤷🏽

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
Inherit doc comments from default implementations of inherited protocols

For example when we have

```swift
protocol BaseProtocol {}

extension BaseProtocol {
  /// Say hello
  func hello() { }
}
struct MyStruct: BaseProtocol {
  func hello() {}
}
```

Then `MyStruct.hello` should inherit the doc comment from the implementation of `BaseProtocol.hello`. Currently no doc comment is associated with `MyStruct.hello`.

rdar://126240546

slavapestov commented on lib/AST/DocComment.cpp:
```diff
@@ -426,24 +426,49 @@ class CommentProviderFinder final {
 
+  /// Check if there is an inherited protocol that has a default implementation
+  /// of `VD` with a doc comment.
   std::optional<ResultWithDecl> findDefaultProvidedDecl(const ValueDecl *VD) {
```
This is just duplicating logic in the witness checker. Default witnesses are recorded in the ProtocolDecl.

See TypeChecker::inferDefaultWitnesses() and ProtocolDecl::getDefaultWitness().

ahoppen replied:
That’s not sufficient though. We aren’t dealing with protocol requirements here, just default methods on a protocol.

For example in the following test, `InheritedProtocol.hello()` needs to inherit the documentation from `BaseProtocol.hello` but there is no protocol requirement that `hello()` can be a witness for. Any suggestion of how to implement that without my implementation?

```swift
protocol BaseProtocol {}
extension BaseProtocol {
  /// Say hello
  func hello() { }
}

protocol InheritedProtocol: BaseProtocol {}
extension InheritedProtocol {
  func hello() { }
}
```

ahoppen commented:
@swift-ci Please smoke test Linux
